### ISSUE TITLE
errors when defining param and type methods on reals

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#12377]
(originally from #12053)

There are problems creating the value version of a `.epsilon` method. The goal is for `.epsilon` to be defined for both real values as well as real types.

For example
``` chapel
proc type (real(64)).epsilon param return 0x1.0p-052:real(64);
proc (real(64)).epsilon param return 0x1.0p-052:real(64);
proc type (real(32)).epsilon param return 0x1.0p-052:real(32);
proc (real(32)).epsilon param return 0x1.0p-052:real(32);

var v64: real(64);
var v32: real(32);
param et64 = real(64).epsilon;
param ev64 = v64.epsilon;
param et32 = real(32).epsilon;
param ev32 = v32.epsilon;
writeln(et64);
writeln(ev64);
writeln(et32);
writeln(ev32);
```

leads to compilation errors currently.

```
zz.chpl:3: In function 'epsilon':
zz.chpl:3: error: 'param' functions cannot return non-'param' values
zz.chpl:10: error: Initializing parameter 'et32' to value not known at compile time
zz.chpl:4: In function 'epsilon':
zz.chpl:4: error: 'param' functions cannot return non-'param' values
zz.chpl:11: error: Initializing parameter 'ev32' to value not known at compile time
```

Note that this may be related to #11646 - or at least the type alias could allow a different way to write it, but runs into other issues.

**Comment by damianmoz** [comment]
The real(32) stuff won't work as a param. See below. But the real(64) should. Try this
```chapel
proc type (real(64)).epsilon param return 0x1.0p-052:real(64);
proc (real(64)).epsilon param return 0x1.0p-052:real(64);
proc type (real(32)).epsilon const return 0x1.0p-023:real(32);
proc (real(32)).epsilon const return 0x1.0p-023:real(32);

var v64: real(64);
var v32: real(32);
param et64 = real(64).epsilon;
param ev64 = v64.epsilon;
const et32 = real(32).epsilon;
const ev32 = v32.epsilon;
writeln(et64);
writeln(ev64);
writeln(et32);
writeln(ev32);
```
This works for me.  I changed the epsilon for real(32) to 2**(-23) to be strictly correct! :)

Note that you still have to fix the problem which demands real(32) be const, not param. I personally think that is is related to #7692 because Chapel thiinks that the hexadecimal real constant is real(64) and the subsequently conversion to real(32) is only resolvable at run-time.

I could be wrong but I think this can be closed and we move back to #7692.

And yes, what is in #7692 is effectively blocking progress.

**Comment by mppf** [comment]
@damianmoz - thanks for the working example. Turns out I was getting confused by a spelling error in the test...

I agree we can close this issue now. Nonetheless I plan to close it when I merge the PR enabling the test in the issue to function.
